pkg/assets/create/creater.go

```diff
@@ -0,0 +1,190 @@
+			return nil, err
+		}
+		return restmapper.NewDiscoveryRESTMapper(gr), nil
+	}
+
+	mapper, err := getRESTMapper()
+	if err != nil {
+		return err
+	}
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var lastCreateError error
+	err = wait.PollImmediateUntil(500*time.Millisecond, func() (bool, error) {
+		err, refresh := create(manifests, client, mapper, options)
+		if err == nil {
+			// No errors means all resources were created successfully, reset lastCreateError and exit the loop.
+			lastCreateError = nil
+			return true, nil
+		}
+		// We got rest-mapper error, force to refresh discovery so next create use updated mapper.
+		if refresh {
+			mapper, err = getRESTMapper()
+			if err != nil {
+				return false, err
+			}
+		}
+		lastCreateError = err
+		return false, nil
+	}, ctx.Done())
+
+	// Return the last observed set of errors from the create process instead of timeout error.
+	if lastCreateError != nil {
+		return lastCreateError
+	}
+
+	return err
+}
+
+func create(manifests map[string]*unstructured.Unstructured, client dynamic.Interface, mapper meta.RESTMapper, options CreateOptions) (error, bool) {
+	// Sort all manifests, so in case they use number prefixes, we follow their order, which will increase
+	// the chances for the create loop to succeed faster.
+	sortedManifestPaths := []string{}
+	for key := range manifests {
+		sortedManifestPaths = append(sortedManifestPaths, key)
+	}
+	sort.Strings(sortedManifestPaths)
+
+	// Record all errors for the given manifest path (so when we report errors, users can see what manifest failed).
+	errs := map[string][]error{}
+
+	// In case we fail to find a rest-mapping for the resource, force to fetch the updated discovery on next run.
+	reloadDiscovery := false
+
+	for _, path := range sortedManifestPaths {
+		gvk := manifests[path].GetObjectKind().GroupVersionKind()
+		mappings, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			errs[path] = append(errs[path], fmt.Errorf("unable to get REST mapping: %v", err))
+			reloadDiscovery = true
+		}
+
+		if mappings.Scope.Name() == meta.RESTScopeNameRoot {
+			_, err = client.Resource(mappings.Resource).Create(manifests[path], metav1.CreateOptions{})
+		} else {
+			_, err = client.Resource(mappings.Resource).Namespace(manifests[path].GetNamespace()).Create(manifests[path], metav1.CreateOptions{})
+		}
+
+		// Resource already exists means we already succeeded
+		// This should never happen as we remove already created items from the manifest list, unless the resource existed beforehand.
+		if kerrors.IsAlreadyExists(err) {
+			continue
+		}
+
+		if err != nil {
+			errs[path] = append(errs[path], fmt.Errorf("failed to create: %v", err))
+			continue
+		}
+
+		// Creation succeeded lets remove the manifest from the list to avoid creating it second time
+		delete(manifests, path)
+	}
+
+	return formatErrors(errs), reloadDiscovery
+}
+
+func formatErrors(errors map[string][]error) error {
```

Comment:
Use [go-multierror][1] instead of rolling our own?

[1]: https://godoc.org/github.com/hashicorp/go-multierror

Reply:
i will have to format each error with prefix when appending it and we will introduce new thing to vendor, is it worth? :-)